src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs

```diff
@@ -2,108 +2,1037 @@
+                if (!foundHeader)
+                {
+                    if (isTargetHeader)
+                    {
+                        if (!string.Equals(targetHeaderValue, foundHeaderValue, StringComparison.OrdinalIgnoreCase))
+                        {
+                            throw new WebSocketException(SR.Format(SR.net_WebSockets_InvalidResponseHeader, targetHeaderName, foundHeaderValue));
+                        }
+                        foundHeader = true;
+                    }
+                }
+                else
+                {
+                    if (isTargetHeader)
+                    {
+                        throw new WebSocketException(SR.Format(SR.net_webstatus_ConnectFailure));
+                    }
+                }
+            }
+
+            private Task<string> ReadLineNoBufferingAsync(Stream stream, CancellationToken cancellationToken)
+            {
+                Task<string> t = ReadLineNoBufferingAsyncPrivate(stream, cancellationToken);
+                return t.Status == TaskStatus.RanToCompletion ? t : WithAbortAsync<string>(t, AbortErrorType.ConnectFailureWebSocketException, cancellationToken);
+            }
+
+            private async Task<string> ReadLineNoBufferingAsyncPrivate(Stream stream, CancellationToken cancellationToken)
+            {
+                // We don't want to use StreamReader.ReadLineAsync or do any other kind of read-ahead buffering, as we don't
+                // want to accidentally consume bytes that are part of subsequent messages.  For now, as this is only used
+                // as part of reading headers, which represents only a small portion of the websocket's lifecycle, we do the
+                // (terribly) inefficient but easy solution of reading one byte at a time.  If this proves problematic, we can
+                // update the implementation to do more buffering, e.g. by having wrapping the underlying stream in a BufferedStream.
+                // As a small concession, we cache the StringBuilder and byte[] used per line.
+
+                KeyValuePair<StringBuilder, byte[]> readLineBuffers = t_readLineBuffers;
+                t_readLineBuffers = default(KeyValuePair<StringBuilder, byte[]>);
+                if (readLineBuffers.Key == null)
+                {
+                    readLineBuffers = new KeyValuePair<StringBuilder, byte[]>(new StringBuilder(), new byte[1]);
+                }
+
+                StringBuilder sb = readLineBuffers.Key;
+                byte[] oneByte = readLineBuffers.Value;
+                Debug.Assert(sb.Length == 0, $"Expected empty StringBuilder");
+                try
+                {
+                    while (true)
+                    {
+                        int numRead = await stream.ReadAsync(oneByte, 0, 1, cancellationToken).ConfigureAwait(false);
+                        byte b = oneByte[0];
+                        if (numRead == 0 || b == '\n') break;
```

Comment:
Since the spec says that newlines are CRLF, can an LF appear within a header?  Do we need to track the two-character state?